python_modules/libraries/dagster-components/dagster_components/core/component_rendering.py

```diff
@@ -51,11 +55,49 @@ def default() -> "TemplatedValueResolver":
     def with_context(self, **additional_context) -> "TemplatedValueResolver":
         return TemplatedValueResolver(context={**self.context, **additional_context})
 
-    def resolve(self, val: Any) -> Any:
+    def _resolve_value(self, val: Any) -> Any:
         return NativeTemplate(val).render(**self.context) if isinstance(val, str) else val
 
+    def _resolve(
+        self,
+        val: Any,
+        valpath: Optional[Sequence[Union[str, int]]],
+        should_render: Callable[[Sequence[Union[str, int]]], bool],
+    ) -> Any:
+        if valpath is not None and not should_render(valpath):
+            return val
+        elif isinstance(val, dict):
+            return {
+                k: self._resolve(v, [*valpath, k] if valpath is not None else None, should_render)
+                for k, v in val.items()
+            }
+        elif isinstance(val, list):
+            return [
+                self._resolve(v, [*valpath, i] if valpath is not None else None, should_render)
+                for i, v in enumerate(val)
+            ]
+        else:
+            return self._resolve_value(val)
 
-def _should_render(
+    def resolve(self, val: Any) -> Any:
+        """Given a raw value, preprocesses it by rendering any templated values."""
+        return self._resolve(val, None, lambda _: True)
+
+    def resolve_params(self, val: T, target_type: Type) -> T:
+        """Given a raw value, preprocesses it by rendering any templated values that are not marked as deferred in the target_type's json schema."""
+        json_schema = (
+            target_type.model_json_schema() if issubclass(target_type, BaseModel) else None
+        )
+        if json_schema is None:
+            should_render = lambda _: True
+        else:
+            should_render = functools.partial(
+                has_required_scope, json_schema=json_schema, subschema=json_schema
+            )
+        return self._resolve(val, [], should_render=should_render)
+
+
+def has_required_scope(
```

Comment:
I think part of my confusion is that this function name is a bit ambigious. Does it mean that there is a field that has the `RenderingScope` attribute?

Reply:
> Does it mean that there is a field that has the RenderingScope attribute?

That's right yeah -- basically should_render is a function that reads the json schema and determines if anything along the path has some required scope

Reply:
`has_rendering_scope`?

Reply:
Still not a fan of this name

Reply:
whoops I forgot to push after the name change